File: bindings/zig/src/safe_cookie.zig
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
//
// Proven SafeCookie - FFI bindings to libproven HTTP cookie operations.
// All computation is performed in verified Idris 2 code via libproven.

const std = @import("std");
const c = @cImport(@cInclude("proven.h"));

/// Error types for cookie operations.
pub const CookieError = error{
    InvalidName,
    InvalidValue,
    ProvenError,
};

/// SameSite attribute values.
pub const SameSite = enum(c_int) {
    strict = c.PROVEN_SAMESITE_STRICT,
    lax = c.PROVEN_SAMESITE_LAX,
    none = c.PROVEN_SAMESITE_NONE,
};

/// Cookie prefix type.
pub const CookiePrefix = enum(i64) {
    none = 0,
    secure = 1,
    host = 2,
};

/// Managed string from libproven.
pub const ProvenString = struct {
    ptr: [*]u8,
    len: usize,

    pub fn slice(self: ProvenString) []const u8 {
        return self.ptr[0..self.len];
    }

    pub fn deinit(self: ProvenString) void {
        c.proven_free_string(self.ptr);
    }
};

/// Check for cookie injection characters via libproven.
pub fn hasInjection(value: []const u8) bool {
    const result = c.proven_cookie_has_injection(value.ptr, value.len);
    return result.status == c.PROVEN_OK and result.value;
}

/// Validate cookie name via libproven.
pub fn validateName(name: []const u8) bool {
    const result = c.proven_cookie_validate_name(name.ptr, name.len);
    return result.status == c.PROVEN_OK and result.value;
}

/// Validate cookie value via libproven.
pub fn validateValue(value: []const u8) bool {
    const result = c.proven_cookie_validate_value(value.ptr, value.len);
    return result.status == c.PROVEN_OK and result.value;
}

/// Get cookie prefix type via libproven.
pub fn getPrefix(name: []const u8) CookiePrefix {
    const result = c.proven_cookie_get_prefix(name.ptr, name.len);
    if (result.status != c.PROVEN_OK) return .none;
    return @enumFromInt(result.value);
}

/// Build Set-Cookie header value via libproven.
/// Caller must call deinit() on the returned ProvenString.
pub fn buildSetCookie(
    name: []const u8,
    value: []const u8,
    domain: ?[]const u8,
    cookie_path: ?[]const u8,
    max_age: i64,
    secure: bool,
    http_only: bool,
    same_site: SameSite,
    partitioned: bool,
) CookieError!ProvenString {
    const attrs = c.ProvenCookieAttributes{
        .domain = if (domain) |d| d.ptr else null,
        .domain_len = if (domain) |d| d.len else 0,
        .path = if (cookie_path) |p| p.ptr else null,
        .path_len = if (cookie_path) |p| p.len else 0,
        .max_age = max_age,
        .secure = secure,
        .http_only = http_only,
        .same_site = @intFromEnum(same_site),
        .partitioned = partitioned,
    };
    const result = c.proven_cookie_build_set_cookie(
        name.ptr,
        name.len,
        value.ptr,
        value.len,
        attrs,
    );
    if (result.status != c.PROVEN_OK) return error.ProvenError;
    return ProvenString{ .ptr = @ptrCast(result.value), .len = result.length };
}

/// Build delete cookie header value via libproven.
/// Caller must call deinit() on the returned ProvenString.
pub fn buildDelete(name: []const u8) CookieError!ProvenString {
    const result = c.proven_cookie_build_delete(name.ptr, name.len);
    if (result.status != c.PROVEN_OK) return error.ProvenError;
    return ProvenString{ .ptr = @ptrCast(result.value), .len = result.length };
}

test "validateName" {
    try std.testing.expect(validateName("session_id"));
}

test "hasInjection" {
    try std.testing.expect(hasInjection("value;with;semicolons"));
    try std.testing.expect(!hasInjection("safe_value"));
}

File: bindings/java/src/main/java/io/github/hyperpolymath/proven/SafeCookie.java
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
package io.github.hyperpolymath.proven;

import com.sun.jna.Memory;

import java.nio.charset.StandardCharsets;
import java.util.Optional;

/**
 * Safe HTTP cookie operations via libproven FFI.
 *
 * Provides cookie injection detection, name/value validation, and prefix detection.
 * All logic delegates to the Idris 2 verified implementation through
 * the Zig C ABI. No cookie logic is reimplemented in Java.
 */
public final class SafeCookie {

    private static final ProvenLibrary LIB = ProvenLibrary.INSTANCE;

    private SafeCookie() {}

    /**
     * Check if a cookie value contains injection sequences.
     *
     * @param value the cookie value to check
     * @return true if injection detected, empty on null pointer error
     */
    public static Optional<Boolean> hasInjection(String value) {
        Memory mem = NativeUtil.toNativeString(value);
        if (mem == null) {
            return Optional.empty();
        }
        byte[] bytes = value.getBytes(StandardCharsets.UTF_8);
        return NativeUtil.extractBool(
            LIB.proven_cookie_has_injection(mem, bytes.length));
    }

    /**
     * Validate a cookie name.
     *
     * @param name the cookie name to validate
     * @return true if valid, empty on null pointer error
     */
    public static Optional<Boolean> validateName(String name) {
        Memory mem = NativeUtil.toNativeString(name);
        if (mem == null) {
            return Optional.empty();
        }
        byte[] bytes = name.getBytes(StandardCharsets.UTF_8);
        return NativeUtil.extractBool(
            LIB.proven_cookie_validate_name(mem, bytes.length));
    }

    /**
     * Validate a cookie value.
     *
     * @param value the cookie value to validate
     * @return true if valid, empty on null pointer error
     */
    public static Optional<Boolean> validateValue(String value) {
        Memory mem = NativeUtil.toNativeString(value);
        if (mem == null) {
            return Optional.empty();
        }
        byte[] bytes = value.getBytes(StandardCharsets.UTF_8);
        return NativeUtil.extractBool(
            LIB.proven_cookie_validate_value(mem, bytes.length));
    }

    /**
     * Detect cookie name prefix (__Secure- or __Host-).
     *
     * @param name the cookie name
     * @return the prefix type as integer code, or empty on error
     */
    public static Optional<Long> getPrefix(String name) {
        Memory mem = NativeUtil.toNativeString(name);
        if (mem == null) {
            return Optional.empty();
        }
        byte[] bytes = name.getBytes(StandardCharsets.UTF_8);
        return NativeUtil.extractLong(
            LIB.proven_cookie_get_prefix(mem, bytes.length));
    }
}
